posts/2019-02-06-algebraic-api-design.md

```diff
@@ -0,0 +1,830 @@
+
+```tut:silent
+object Game {
+  // let's follow the naming convention used by Cats and call this 'laws'
+  object laws {
+    import cats.Functor
+    import cats.implicits._
+
+    def simulationIsTerminal[F[_]: Functor, Move, BoardPosition, Score](
+        gameState: GameState[Move, BoardPosition, Score])(
+        implicit ev: Game[F, Move, BoardPosition, Score]): F[Boolean] =
+      ev.simulation(gameState).map(ev.legalMoves).map(_.isEmpty)
+
+    def legalMoveModifiesGameState[F[_], Move, BoardPosition, Score](
+        gameState: GameState[Move, BoardPosition, Score],
+        move: Move)(implicit ev: Game[F, Move, BoardPosition, Score]): Boolean = {
+      val legalMoves    = ev.legalMoves(gameState)
+      val nextGameState = ev.applyMove(gameState, move)
+      !legalMoves.contains(move) ||
+        (nextGameState.position != gameState.position && nextGameState.playedMoves.length == gameState.playedMoves.length + 1)
+    }
+  }
+}
+```
+
+There is an additional constraint that we discover while implementing the predicates that define the properties of `Game`. `F` must have a `Functor` instance. `Functor` is the least powerful abstraction that allows us to access the value inside `F` and express the property related to `simulation`. In the next section we will see that the program we are going to write will imply additional constraints for `F`.
+
+The properties of `Game` are tightly coupled to the algebra as any implementation must satisfy them. This is why they are defined inside the companion object of `Game` as part of the library code.
+
+Later we will see how to implement property-based tests to verify the properties.
+
+### Programs
+
+The `Game` properties are expressed solely in terms of the `Game` algebra. We have no idea how `Game` is implemented or what the types `Move`, `BoardPosition`, or `Score` look like. In a similar fashion we will only use the algebra to implement programs such as the search algorithm described above.
+
+Before we do this, we will define another algebra that describes logging, simply for convenience. Especially during long running searches it is useful to be able to output intermediate results and search states:
+
+```tut:silent
+import cats.Show
+
+trait Logger[F[_]] {
+  def log[T: Show](t: T): F[Unit]
+}
+
+object Logger {
+  def apply[F[_]]()(implicit ev: Logger[F]): Logger[F] = ev
+}
+```
+
+With the two algebras `Game` and `Logger` we can now implement the nested Monte Carlo tree search.
+
+```tut:book
+import cats.Monad
+import cats.implicits._
+
+def nestedSearch[F[_]: Monad: Logger, Move, Position, Score](
+    numLevels: Int,
+    level: Int,
+    gameState: GameState[Move, Position, Score])(
+    implicit g: Game[F, Move, Position, Score],
+    ord: Ordering[Score],
+    show: Show[GameState[Move, Position, Score]]): F[GameState[Move, Position, Score]] = {
+  val legalMoves = g.legalMoves(gameState)
+  for {
+    _ <- if (level == numLevels) Logger[F].log(gameState) else Monad[F].pure(())
```

Comment:
I'd recommend using syntax: `().pure[F]` or `Monad[F].unit` if you don't like syntax. Same for the case below.